packages/browser-extension/src/hooks/useSteps.ts

```diff
@@ -107,40 +171,58 @@ export const calculateSteps = ({
   history: BrowsingHistoryItem[];
   isZkProvingDone: boolean;
 }) => {
-  return stepsSetup.reduce((accumulator, currentStep) => {
+  const steps: Step[] = [];
+
+  for (const currentStep of stepsSetup) {
     const hasUncompletedStep =
-      accumulator.length > 0 &&
-      accumulator[accumulator.length - 1]?.status !== StepStatus.Completed;
+      steps.length > 0 && steps.at(-1)?.status !== StepStatus.Completed;
 
-    const mappedStep = {
+    steps.push({
+      step: currentStep,
       label: currentStep.label,
       link: currentStep.url,
       kind: currentStep.step,
-      status: calculateStepStatus({
+      status: await calculateStepStatus({
         hasUncompletedStep,
         step: currentStep,
         history,
         isZkProvingDone,
       }),
-    };
-    return [...accumulator, mappedStep];
-  }, [] as Step[]);
+    });
+  }
+
+  return steps;
 };
 
 export const useSteps = (): Step[] => {
   const [config] = useProvingSessionConfig();
   const [history] = useBrowsingHistory();
   const { isDone: isZkProvingDone } = useZkProvingState();
+  const [steps, setSteps] = useState<Step[]>([]);
 
   const stepsSetup = match(config)
     .with(LOADING, () => [])
     .with(P.nullish, () => [])
     .with({ steps: P.array(P.any) }, ({ steps }) => steps)
     .exhaustive();
 
-  return calculateSteps({
-    stepsSetup,
-    history,
-    isZkProvingDone,
-  });
+  useEffect(() => {
+    let isCancelled = false;
+    void (async () => {
+      const steps = await calculateSteps({
+        stepsSetup,
+        history,
+        isZkProvingDone,
+      });
+      if (!isCancelled) {
+        setSteps(steps);
+      }
+    })();
+
+    return () => {
+      isCancelled = true;
+    };
+  }, [stepsSetup, history, isZkProvingDone]);
```

Comment:
Its suspicious,  based on dependencies this wont change stel status even though element we depend on in 'useAction` step will appear. There should be synchronisation between extension state and the DOM. Otherwise if elelement we expect will appear with some delay ( animation ? ) we will stuck. this should be added as test case.